[c10d] Fix timeout dump path write path overlap when there are multiple PGs

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116316
* #116520
* __->__ #116218

Basically we observed that if there are multiple PGs and if the timeout happens on one of the subPG, we somehow use the local rank in the dump file. We realize that:
1. For setting the timeout signal in the store, any watchdog thread from any PG can do that.
2. For checking and dump, only the watchdog thread of default PG which we will always create and contain all ranks (no file name conflict) is needed here because the store signal and dump debug info are all global.
3. Since dump is global, we want to avoid the case when ranks from sub-PG pollute logs from global ranks (local rank 0 vs global rank 0). So that we use global ranks here to initialize debug info writer. (Down the road, we are thinking about making it a singleton so that user only register it once for multi-PG case.) 

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @wz337 @tianyu-l @wconstab @yf225

fduwjj committed:
[c10d] Fix timeout dump when there are multiple PGs

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1440,10 +1440,12 @@ void ProcessGroupNCCL::watchdogHandler() {
 
-    // poll store to see if some ranks have flagged a timeout when
+    // Assuming that we always init a process group containing all ranks,
+    // we only use the watchdog thread for that PG to detect timeout dump.
```
> to detect timeout dump

change to
"to listen for the global signal to dump and abort"

because we still are using every thread for locally detecting timeout/dump, and this might confuse ppl

fduwjj replied:
ok..

fduwjj committed:
Update on "[c10d] Fix timeout dump when there are multiple PGs"


Basically we observed that if there are multiple PGs and if the timeout happens on one of the subPG, we somehow use the local rank in the dump file. We realize that:
1. For setting the timeout signal in the store, any watchdog thread from any PG can do that.
2. For checking and dump, only the watchdog thread of default PG which we will always create and contain all ranks (no file name conflict) is needed here because the store signal and dump debug info are all global.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Fix timeout dump when there are multiple PGs"


Basically we observed that if there are multiple PGs and if the timeout happens on one of the subPG, we somehow use the local rank in the dump file. We realize that:
1. For setting the timeout signal in the store, any watchdog thread from any PG can do that.
2. For checking and dump, only the watchdog thread of default PG which we will always create and contain all ranks (no file name conflict) is needed here because the store signal and dump debug info are all global.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Fix timeout dump when there are multiple PGs"


Basically we observed that if there are multiple PGs and if the timeout happens on one of the subPG, we somehow use the local rank in the dump file. We realize that:
1. For setting the timeout signal in the store, any watchdog thread from any PG can do that.
2. For checking and dump, only the watchdog thread of default PG which we will always create and contain all ranks (no file name conflict) is needed here because the store signal and dump debug info are all global.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Fix timeout dump when there are multiple PGs"


Basically we observed that if there are multiple PGs and if the timeout happens on one of the subPG, we somehow use the local rank in the dump file. We realize that:
1. For setting the timeout signal in the store, any watchdog thread from any PG can do that.
2. For checking and dump, only the watchdog thread of default PG which we will always create and contain all ranks (no file name conflict) is needed here because the store signal and dump debug info are all global.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/TraceUtils.h:
```diff
@@ -269,10 +269,10 @@ inline std::string retrieveDesyncReport(
 #ifdef USE_C10D_NCCL
 
-DebugInfoWriter::DebugInfoWriter(int rank) {
+DebugInfoWriter::DebugInfoWriter(int pgId, int rank) {
```
i don't think we should need pgId in DebugInfoWriter.  The whole idea was the debug info is process-wide, so its not bound to a specific PG.  

If you're just using it here to log and ensure we only use pg 0 then i think thats ok.  but no need to pollute the filename with an extra 0_ IMO

fduwjj replied:
Well, actually during timeout, subPG will use this to dump the debug info and it will use local rank here. Let's say a local rank 0 (global rank 8) will write debug info when timeout in watchdog, then it will write into file name "*_rank_0" while for the global rank 0, it will also dump debug info to the file name "*_rank_0" and this is a race which will mess up our final dump files.

fduwjj replied:
For example in the log I see log files:

<img width="103" alt="image" src="https://github.com/pytorch/pytorch/assets/6937752/adcf68e5-9d3c-4bd8-8e3d-35f0af673092">

They are all from the same host, if we don't have uid in the file name, the file will all be the same.

fduwjj replied:
Hmm actually different sub-PG shared the same uid_, I will use global rank here instead.

fduwjj committed:
Update on "[c10d] Fix timeout dump when there are multiple PGs"


Basically we observed that if there are multiple PGs and if the timeout happens on one of the subPG, we somehow use the local rank in the dump file. We realize that:
1. For setting the timeout signal in the store, any watchdog thread from any PG can do that.
2. For checking and dump, only the watchdog thread of default PG which we will always create and contain all ranks (no file name conflict) is needed here because the store signal and dump debug info are all global.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Fix timeout dump when there are multiple PGs"


Basically we observed that if there are multiple PGs and if the timeout happens on one of the subPG, we somehow use the local rank in the dump file. We realize that:
1. For setting the timeout signal in the store, any watchdog thread from any PG can do that.
2. For checking and dump, only the watchdog thread of default PG which we will always create and contain all ranks (no file name conflict) is needed here because the store signal and dump debug info are all global.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Fix timeout dump path write path overlap when there are multiple PGs"


Basically we observed that if there are multiple PGs and if the timeout happens on one of the subPG, we somehow use the local rank in the dump file. We realize that:
1. For setting the timeout signal in the store, any watchdog thread from any PG can do that.
2. For checking and dump, only the watchdog thread of default PG which we will always create and contain all ranks (no file name conflict) is needed here because the store signal and dump debug info are all global.
3. Since dump is global, we want to avoid the case when ranks from sub-PG pollute logs from global ranks (local rank 0 vs global rank 0). So that we use global ranks here to initialize debug info writer. (Down the road, we are thinking about making it a singleton so that user only register it once for multi-PG case.) 

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge
